lib/pos_blockchain.go

```diff
@@ -57,66 +51,179 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
+	if exists {
+		if blockNode.IsValidateFailed() {
+			return blockNode, errors.New("isValidPoSBlock: Block already marked as ValidateFailed")
+		}
+		if blockNode.IsValidated() {
+			return blockNode, nil
+		}
+	}
 	// Make sure its parent is validated. If not, we need to validate it first by calling processBlockPoS.
 	// Note that if any other ancestor is not validated, calling processBlockPoS on the parent will
 	// end up calling processBlockPoS on all of its ancestors that aren't validated yet.
 	parentBlockNode, exists := bc.blockIndex[*block.Header.PrevBlockHash]
 	if !exists {
-		return false, false, nil, errors.New("processBlockPoS: Parent block not found in block index. This should never happen.")
+		// We should never really hit this, but if we do, we upsert the block w/ StatusBlockStored
+		// but not StatusBlockValidateFailed. This makes sure this function is idempotent.
+		blockNode, err = bc.storeBlockInBlockIndex(block)
+		if err != nil {
+			return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem adding block to block index: ")
+		}
+		return blockNode, errors.New("isValidPoSBlock: Parent block not found in block index. This should never happen.")
 	}
 	if parentBlockNode.IsValidateFailed() {
 		// We'll never add this to the best chain because its parent is invalid.
 		// If its parent is ValidateFailed, we can update the block index with the block status ValidateFailed
-		if err = bc.storeValidateFailedBlockInBlockIndex(block); err != nil {
-			return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem adding validate failed block to block index")
+		blockNode, err = bc.storeValidateFailedBlockInBlockIndex(block)
+		if err != nil {
+			return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem adding validate failed block to block index")
 		}
-		return false, false, nil, errors.New("processBlockPoS: block's parent failed validation: ")
+		return blockNode, errors.New("isValidPoSBlock: block's parent failed validation: ")
 	}
 	if !parentBlockNode.IsValidated() {
-		blk, err := GetBlock(parentBlockNode.Hash, bc.db, bc.snapshot)
+		var parentBlock *MsgDeSoBlock
+		parentBlock, err = GetBlock(parentBlockNode.Hash, bc.db, bc.snapshot)
 		if err != nil {
-			return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem getting parent block %v", parentBlockNode.Hash)
-		}
-		_, _, _, err = bc.processBlockPoS(blk, currentView, verifySignatures)
-		if err != nil {
-			return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem processing parent block %v", parentBlockNode.Hash)
+			// If we don't have the parent, we need to fetch it from the network.
+			// We won't handle that in here, so we should just store the block and return.
+			blockNode, storeInBlockIndexErr = bc.storeBlockInBlockIndex(block)
+			if storeInBlockIndexErr != nil {
+				return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding block to block index: %v", storeInBlockIndexErr)
```

Comment:
The algorithm failed here too. The resulting blockNode isn't trustworthy, so we shouldn't return it
```suggestion
				return nil, errors.Wrapf(err, "isValidPoSBlock: Problem adding block to block index: %v", storeInBlockIndexErr)
```